CDAP-7702 : Adding generation id as part of rowkey of TMS Data tables

gokulavasan opened the pull request:
TopicIds should have generations ( 1...MAXINT) so that when a topic is deleted and recreated, the messages present in the previous generation should not be considered during fetching. 

This is achieved by having a generation id as part of the rowKey in TMS Data tables. TopicMetada in MetadataTable keeps track of the last used generation id. This implies that rows are never deleted from MetadataTable. Ideally, we could delete rows once all the data from Data Tables corresponding to that topic has been removed, but that is not part of this PR.

When a topic is deleted, the generation id in MetadataTable for that TopicId, is made -ve of the current generation id. When the topic is recreated again, then the generation id is made +ve and it is bumped up by 1.

JIRA : https://issues.cask.co/browse/CDAP-7702
Build : http://builds.cask.co/browse/CDAP-RUT299-11

gokulavasan committed:
Introducing generation id for TMS Topics

Marking deletion in MessageTable instead of deleting, and remove the ability to delete in PayloadTable

Store Generation id as part of the properties

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/TopicMetadata.java:
```diff
@@ -56,6 +57,14 @@ public TopicMetadata(TopicId topicId, Map<String, String> properties, boolean va
+  public TopicMetadata(TopicId topicId, Map<String, String> properties, int generation) {
+    this.topicId = topicId;
+    this.properties = ImmutableMap.copyOf(properties);
+    this.generation = generation;
```
Why the one with generation doesn't validate? Because it always reading from storage, hence should already been validated?

gokulavasan replied:
I will add a check to make sure that the generation is never zero.

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/TopicMetadata.java:
```diff
@@ -84,6 +93,20 @@ public TopicId getTopicId() {
+  /**
+   * Check whether the topic has been marked as deleted or not.
+   */
+  public boolean isDeleted() {
```
Better call it `exists` (and flip the condition) to align with many other classes in CDAP and in Java.

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/service/MessageTableStoreRequestWriter.java:
```diff
@@ -35,17 +36,18 @@
 
   @Override
-  MessageTable.Entry getEntry(TopicId topicId, boolean transactional, long transactionWritePointer,
+  MessageTable.Entry getEntry(TopicId topicId, int generation, boolean transactional, long transactionWritePointer,
```
Seems like you can just change the signature to take `TopicMetadata`.

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/MessagingUtils.java:
```diff
@@ -30,18 +30,30 @@ private MessagingUtils() {
+   * @return byte array representation to be used as row key prefix for data tables
+   */
+  public static byte[] toDataKeyPrefix(TopicId topicId, int generation) {
+    return Bytes.add(toMetadataRowKey(topicId), Bytes.toBytes(generation));
```
This would generate three `byte[]`. Create one that fits and use `Bytes.putInt()`.

gokulavasan replied:
I will make that change. Guess it will still result in two ```byte[]``` creation?

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/service/StoreRequestWriter.java:
```diff
@@ -59,7 +61,8 @@
    *                                 a {@link PendingStoreRequest} has an empty iterator of payload.
    */
-  StoreRequestWriter(TimeProvider timeProvider, boolean generateNullPayloadEntry) {
+  StoreRequestWriter(TopicMetadata metadata, TimeProvider timeProvider, boolean generateNullPayloadEntry) {
```
This is not entirely right. Each individual `PendingStoreRequest` can potentially publish to different `TopicId`, hence the `TopicMetadata` would be different. The current implementation in `CoreMessagingService` does use one `StoreRequestWriter` instance per topic, but it is not guaranteed in future. From this class perspective, The `PendingStoreRequest` should contains `TopicMetadata` so that generation id associated with the corresponding topic is available at write time.

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/store/TopicMetadataCodec.java:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Codec class to serialize/deserialize {@link TopicMetadata}.
+ */
+public class TopicMetadataCodec implements JsonSerializer<TopicMetadata>, JsonDeserializer<TopicMetadata> {
```
Why need a codec? you can mark fields that you don't want to be serialized as `transient`.

gokulavasan replied:
We used to just persist the map of properties before. But now, we need to persist the generation as well. Is there a better way to do it other than using a codec?

chtyim replied:
`TopicMetadata` only has simple fields. We can just mark fields that we don't want to serialize as `transient` (e.g. `validated`).

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMetadataTable.java:
```diff
@@ -77,35 +74,62 @@ public TopicMetadata getMetadata(TopicId topicId) throws IOException, TopicNotFo
+    byte[] value = result.getValue(columnFamily, COL);
+    if (value == null) {
+      putBuilder.add(columnFamily, COL, Bytes.toBytes(GSON.toJson(topicMetadata)));
+      if (!hTable.checkAndPut(rowKey, columnFamily, COL, null, putBuilder.build())) {
```
This is not entirely right. It can also be a topic was created and then deleted (from other threads) in between the `get` call and this `checkAndPut` call in this method. You have do a while loop.

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMetadataTable.java:
```diff
@@ -77,35 +74,62 @@ public TopicMetadata getMetadata(TopicId topicId) throws IOException, TopicNotFo
+      TopicMetadata newMetadata = new TopicMetadata(metadata.getTopicId(), metadata.getProperties(), newGenerationId);
+      byte[] newValue = Bytes.toBytes(GSON.toJson(newMetadata));
+      putBuilder.add(columnFamily, COL, newValue);
+      if (!hTable.checkAndPut(rowKey, columnFamily, COL, value, putBuilder.build())) {
```
Similar to my comment above.

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMetadataTable.java:
```diff
@@ -77,35 +74,62 @@ public TopicMetadata getMetadata(TopicId topicId) throws IOException, TopicNotFo
-      TopicMetadata oldMetadata = getMetadata(topicMetadata.getTopicId());
-      byte[] oldValue = Bytes.toBytes(GSON.toJson(new TreeMap<>(oldMetadata.getProperties()), MAP_TYPE));
+      oldMetadata = getMetadata(topicMetadata.getTopicId());
+      byte[] oldValue = Bytes.toBytes(GSON.toJson(oldMetadata));
```
You should also stop the loop and throw topic exists exception if the `oldMetadata` is marked as deleted (-ve generation).

gokulavasan replied:
That is already done in the getMetadata call. It will throw TopicNotFoundException if the generation is negative (ie, exists call returns false)

gokulavasan commented:
@chtyim Addressed comments. Have a question about the Codec comment. Please take a look again. Thanks!

gokulavasan commented:
@chtyim Removed the Codec class. PR is ready for another review. Thanks!

gokulavasan commented:
@chtyim The last commit (5fb7dc8) contains the changes to mark deletion of entries in MessageTable instead of deleting them from Message and Payload Tables. I have included it in this PR since it has minor changes but I can move it out to a separate PR if required.

gokulavasan commented:
@chtyim Rebased against release/4.0 and resolved conflicts. Please take a look when you get a chance.

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/client/StoreRequestBuilder.java:
```diff
@@ -44,7 +44,7 @@
    * Creates a new {@link StoreRequestBuilder} instance.
    *
-   * @param topicId the topic that the store action will happen.
+   * @param topicId metadata of the topic id {@link TopicId}.
```
I think the javadoc shouldn't get changed.

gokulavasan replied:
Ah yes, didn't resolve that conflict properly.

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/service/CoreMessagingService.java:
```diff
@@ -118,6 +118,8 @@ public void updateTopic(TopicMetadata topicMetadata) throws TopicNotFoundExcepti
       metadataTable.deleteTopic(topicId);
+      messageTableWriterCache.invalidate(topicId);
+      payloadTableWriterCache.invalidate(topicId);
       topicCache.invalidate(topicId);
```
Should invalid the `topicCache` first.

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/service/CoreMessagingService.java:
```diff
@@ -186,20 +190,12 @@ public void rollback(TopicId topicId, RollbackDetail rollbackDetail) throws Topi
       // Safe to cast to short because message table only use the sequence id as unsigned bytes.
-      messageTable.delete(topicId, rollbackDetail.getStartTimestamp(), (short) rollbackDetail.getStartSequenceId(),
-                          rollbackDetail.getEndTimestamp(), (short) rollbackDetail.getEndSequenceId());
+      messageTable.markDeleted(topicId, metadata.getGeneration(), rollbackDetail.getStartTimestamp(),
```
Better to call the method `undo` or `rollback`. How it is done is internal to the MessageTable.

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/store/AbstractMessageTable.java:
```diff
@@ -117,13 +117,26 @@ public void delete(TopicId topicId, long startTimestamp, short startSequenceId,
+    return Bytes.toLong(txPtr);
+  }
+
+  protected byte[] encodeTx(long txPtr) {
```
Do we need these two one line methods? Seems pretty straightforward to just call `Bytes.` directly.

gokulavasan replied:
@chtyim I added these two methods so that LevelDB, HBase implementations can use it without knowing how encoding, decoding happens.

chtyim replied:
Having utility like method in parent is not a very good idea, as it doesn't govern the behavior (e.g. when this method is used for what purpose), which the parent should do. If we need util method, It's better to have one method call `byte[] flipSign(byte[] longBytes)` instead, which we can put in `Bytes` util class. And flipping doesn't need to decode it to long.

chtyim replied:
It can be easily done with this:

```java
    int inc = 1;
    for (int i = bytes.length - 1; i >= 0; i--) {
      int val = (bytes[i] ^ 0xff) + inc;
      bytes[i] = (byte) val;
      if (val <= 0xff) {
        inc = 0;
      }
    }
```

gokulavasan replied:
After passing in txBytes, I removed those two methods. So we don't need to do this now. But I agree that the flipSign is a useful method to have in Bytes class!

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/store/MessageTable.java:
```diff
@@ -86,7 +91,7 @@
    * @return a {@link CloseableIterator} of {@link Entry}.
    */
-  CloseableIterator<Entry> fetch(TopicId topicId, long startTime,
+  CloseableIterator<Entry> fetch(TopicId topicId, int generation, long startTime,
```
Shall we take `TopicMetadata` instead of `TopicId` and `int generation`? Is there a case that `TopicMetadata` is not available when this method is called?

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMessageTable.java:
```diff
@@ -130,20 +128,30 @@ public void delete(byte[] startKey, byte[] stopKey) throws IOException {
-        batchDeletes.add(tableUtil.buildDelete(result.getRow()).build());
+        // No need to turn the key back to the original row key because we want to put with the actual row key
+        PutBuilder putBuilder = tableUtil.buildPut(result.getRow());
+        byte[] payload = result.getValue(columnFamily, PAYLOAD_COL);
```
Why we have to care about the payload? I think we can just scan without the payload column, but with the `t` column only and only update the `t` column

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMessageTable.java:
```diff
@@ -130,20 +128,30 @@ public void delete(byte[] startKey, byte[] stopKey) throws IOException {
+        }
+
+        byte[] txCol = result.getValue(columnFamily, TX_COL);
+        if (txCol != null) {
```
It shouldn't be null. In fact the `txId` that this column decoded to should match with the `txId` as given from the `RollbackDetail` (currently not passed to this method, but it is available from the caller of this method, so instead of taking four parameters, we can just take `RollbackDetail`).

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMetadataTable.java:
```diff
@@ -77,51 +70,81 @@ public TopicMetadata getMetadata(TopicId topicId) throws IOException, TopicNotFo
 
   @Override
   public void createTopic(TopicMetadata topicMetadata) throws TopicAlreadyExistsException, IOException {
+    boolean completed = false;
```
Move this closer to the `while` loop, which is where it is getting used.

gokulavasan commented:
@chtyim I have addressed comments (https://github.com/caskdata/cdap/pull/7238/commits/51082af3ead566d3cdc465c78f3fcbb91f856869) and I have also made the change to store generation id as part of the properties (https://github.com/caskdata/cdap/pull/7238/commits/2d486c138d8321d1d144847d351c2a0d68c5297a). Please take a look when you get a chance. Thanks!

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/TopicMetadata.java:
```diff
@@ -30,10 +30,11 @@
+  public static final String GENERATION_KEY = "generation";
 
-  private final TopicId topicId;
+  private final transient TopicId topicId;
```
We actually don't need to add `transient` now because we are not serializing this class, right?

chtyim commented:
Just one comment left, the rest LGTM

gokulavasan commented:
Thanks for the review @chtyim. I have addressed your last comment. I will squash and merge after the final build passes.
